### PR TITLE
executor: add service discovery in executor server

### PR DESCRIPTION
**Pull request by amyangfei**
Finish task-3 in #91
- Rename `keepalive` to `discoveryKeepalive` in executor server, and refactor this function.
   - When meeting session done or watch error from discovery service, doesn't return an error and terminate executor, retry to connect to discovery service(meta store) instead. We do this because the aliveness of executor is decided by the **heartbeat loop**, the discovery keepalive loop is only used to maintain executor address information.
- Add unit test to `discoveryKeepalive`
- Integrate `MessageRouter` interface in executor server.

### CODE REVIEW

File: executor/server.go [hunk 1]
```diff
@@ -50,6 +62,7 @@ func NewServer(cfg *Config, ctx *test.Context) *Server {
 		testCtx:     ctx,
 		cliUpdateCh: make(chan []string),
 	}
+	s.discoveryConnector = s.connectToEtcdDiscovery
```
Comment (hanfei1991) [inline executor/server.go:65]: Why delegate this function instead of using it directly?

Comment (amyangfei) [reply executor/server.go:65]: Use a function variable here to make it easy for unit test.

**Review by hanfei1991 — APPROVED** [review]

File: executor/server.go [hunk 2]
```diff
@@ -131,6 +144,15 @@ func (s *Server) Stop() {
 		}
 	}
 
+	if s.metastore != nil {
```
Comment (liuzix) [inline executor/server.go:147]: Can you add a comment why it is necessary to call `Delete` here?

File: executor/server.go [hunk 3]
```diff
@@ -260,29 +282,96 @@ func (s *Server) keepalive(ctx context.Context) error {
...
+	if err != nil {
+		return err
+	}
+	executors, err := s.discovery.Snapshot(ctx)
 	if err != nil {
 		return err
 	}
+	for uuid, exec := range executors {
+		if uuid == string(s.info.ID) {
+			continue
```
Comment (liuzix) [inline executor/server.go:338]: I think we should not skip adding itself to the router.

File: pkg/srvdiscovery/interface.go [hunk 4]
```diff
@@ -17,18 +17,23 @@ type ServiceResource = model.ExecutorInfo
...
 
 	// Watch watches the change of service resources, the watched events will be
 	// returned through a channel.
 	Watch(ctx context.Context) <-chan WatchResp
+
+	// SnapshotClone returns the clone of cached snapshot in Discovery
+	SnapshotClone() (map[UUID]ServiceResource, Revision)
+
+	// CopySnapshot copies snapshot from given Discovery
+	CopySnapshot(snapshot map[UUID]ServiceResource, revision Revision)
```
Comment (liuzix) [inline pkg/srvdiscovery/interface.go:38]: Probably we should comment on when `CopySnapshot` is useful?

File: executor/server.go (hunk 3 above)
Comment (amyangfei) [reply executor/server.go]: addressed

File: executor/server.go (hunk 2 above)
Comment (amyangfei) [reply executor/server.go:147]: added comment

File: pkg/srvdiscovery/interface.go (hunk 4 above)
Comment (amyangfei) [reply pkg/srvdiscovery/interface.go:42]: added comment

**Comment by amyangfei** [comment]
PTAL @liuzix

**Review by liuzix — APPROVED** [review]

**Pull request merged**